services/searchengine/bleveengine/bleve.go

```diff
@@ -0,0 +1,270 @@
+
+	return indexMapping
+}
+
+func createOrOpenIndex(cfg *model.Config, indexName string, mapping *mapping.IndexMappingImpl) (bleve.Index, error) {
+	indexPath := filepath.Join(*cfg.BleveSettings.IndexDir, indexName+".bleve")
+	if index, err := bleve.Open(indexPath); err == nil {
+		return index, nil
+	}
+
+	index, err := bleve.New(indexPath, mapping)
+	if err != nil {
+		return nil, err
+	}
+	return index, nil
+}
+
+func NewBleveEngine(cfg *model.Config, jobServer *jobs.JobServer) *BleveEngine {
+	return &BleveEngine{
+		cfg:       cfg,
+		jobServer: jobServer,
+	}
+}
+
+func (b *BleveEngine) openIndexes() *model.AppError {
+	if atomic.LoadInt32(&b.ready) != 0 {
+		return model.NewAppError("Bleveengine.Start", "bleveengine.already_started.error", nil, "", http.StatusInternalServerError)
+	}
+
+	var err error
+	b.PostIndex, err = createOrOpenIndex(b.cfg, "posts", getPostIndexMapping())
+	if err != nil {
+		return model.NewAppError("Bleveengine.Start", "bleveengine.create_post_index.error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	b.UserIndex, err = createOrOpenIndex(b.cfg, "users", getUserIndexMapping())
+	if err != nil {
+		return model.NewAppError("Bleveengine.Start", "bleveengine.create_user_index.error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	b.ChannelIndex, err = createOrOpenIndex(b.cfg, "channels", getChannelIndexMapping())
+	if err != nil {
+		return model.NewAppError("Bleveengine.Start", "bleveengine.create_channel_index.error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	atomic.StoreInt32(&b.ready, 1)
+	return nil
+}
+
+func (b *BleveEngine) Start() *model.AppError {
+	if !*b.cfg.BleveSettings.EnableIndexing || *b.cfg.BleveSettings.IndexDir == "" {
+		return nil
+	}
+
+	b.Mutex.Lock()
+	defer b.Mutex.Unlock()
+
+	mlog.Info("EXPERIMENTAL: Starting Bleve")
+
+	return b.openIndexes()
+}
+
+func (b *BleveEngine) closeIndexes() *model.AppError {
+	if b.IsActive() {
+		if err := b.PostIndex.Close(); err != nil {
+			return model.NewAppError("Bleveengine.Stop", "bleveengine.stop_post_index.error", nil, err.Error(), http.StatusInternalServerError)
+		}
+
+		if err := b.UserIndex.Close(); err != nil {
+			return model.NewAppError("Bleveengine.Stop", "bleveengine.stop_user_index.error", nil, err.Error(), http.StatusInternalServerError)
+		}
+
+		if err := b.ChannelIndex.Close(); err != nil {
+			return model.NewAppError("Bleveengine.Stop", "bleveengine.stop_channel_index.error", nil, err.Error(), http.StatusInternalServerError)
+		}
+	}
+
+	atomic.StoreInt32(&b.ready, 0)
+	return nil
+}
+
+func (b *BleveEngine) Stop() *model.AppError {
+	b.Mutex.Lock()
+	defer b.Mutex.Unlock()
+
+	mlog.Info("Stopping Bleve")
+
+	return b.closeIndexes()
+}
+
+func (b *BleveEngine) IsActive() bool {
+	return atomic.LoadInt32(&b.ready) == 1
+}
+
+func (b *BleveEngine) IsIndexingSync() bool {
+	return b.indexSync
+}
+
+func (b *BleveEngine) RefreshIndexes() *model.AppError {
+	return nil
```

Comment:
I'm more inclined to let the method do nothing in case that we trigger the method in a generic way (like, index something & refresh indexes right away). If this specific implementation doesn't need to refresh indexes, then we just do nothing. If we fail, we're going to start distinguishing which interface needs to do what or controlling specific "Not implemented" errors, adding complexity to the logic